src/dlgRoomProperties.cpp

```diff
@@ -0,0 +1,555 @@
+        qsl("background-color: %1; color: %2; border: 1px solid; border-radius: 1px;")
+            .arg(realSymbolColor.name(), backgroundBasedColor(realSymbolColor).name()));
+    }
+    label_preview->setFont(getFontForPreview(newSymbol));
+    label_preview->setText(newSymbol);
+    label_preview->setStyleSheet(
+        qsl("color: %1; background-color: %2; border: %3;")
+            .arg(realSymbolColor.name(), mRoomColor.name(), mpHost->mMapperShowRoomBorders ? qsl("1px solid %1").arg(mpHost->mRoomBorderColor.name()) : qsl("none")));
+
+}
+
+
+QFont dlgRoomProperties::getFontForPreview(QString symbolString)
+{
+    auto font = mpHost->mpMap->mMapSymbolFont;
+    font.setPointSize(font.pointSize() * 0.9);
+    if (!symbolString.isEmpty()) {
+        QFontMetrics mapSymbolFontMetrics = QFontMetrics(font);
+        QVector<quint32> codePoints = symbolString.toUcs4();
+        QVector<bool> isUsable;
+        for (int i = 0; i < codePoints.size(); ++i) {
+            isUsable.append(mapSymbolFontMetrics.inFontUcs4(codePoints.at(i)));
+        }
+        bool needToFallback = isUsable.contains(false);
+        if (needToFallback) {
+            symbolString = QString(QChar::ReplacementCharacter);
+            font.setStyleStrategy(static_cast<QFont::StyleStrategy>(mpHost->mpMap->mMapSymbolFont.styleStrategy() & ~(QFont::NoFontMerging)));
+        }
+    }
+    return font;
+}
+
+
+void dlgRoomProperties::slot_openSymbolColorSelector()
+{
+    auto* dialog = (selectedSymbolColor != nullptr) ? new QColorDialog(selectedSymbolColor, this) : new QColorDialog(defaultSymbolColor(), this);
+    dialog->setAttribute(Qt::WA_DeleteOnClose);
+    dialog->setWindowTitle(tr("Pick color"));
+    dialog->open(this, SLOT(slot_symbolColorSelected(const QColor&)));
+}
+
+
+void dlgRoomProperties::slot_symbolColorSelected(const QColor& color)
+{
+    selectedSymbolColor = color;
+    slot_updatePreview();
+}
+
+
+void dlgRoomProperties::slot_resetSymbolColor()
+{
+    selectedSymbolColor = QColor();
+    slot_updatePreview();
+}
+
+
+QColor dlgRoomProperties::backgroundBasedColor(QColor background)
+{
+    return background.lightness() > 127 ? Qt::black : Qt::white;
+}
+
+QColor dlgRoomProperties::defaultSymbolColor()
+{
+    return backgroundBasedColor(mRoomColor);
+}
+
+
+void dlgRoomProperties::slot_selectRoomColor(QListWidgetItem* pI)
+{
+    mRoomColorNumber = pI->text().toInt();
+}
+
+
+void dlgRoomProperties::slot_defineNewColor()
+{
+    auto color = QColorDialog::getColor(mpHost->mRed, this);
+    if (color.isValid()) {
+        auto environmentId = mpHost->mpMap->mCustomEnvColors.size() + 257 + 16;
+        if (mpHost->mpMap->mCustomEnvColors.contains(environmentId)) {
+            // find a new environment ID to use, starting with the latest
+            // 'safe' number so the new environment is last in the dialog
+            do {
+                environmentId++;
+            } while (mpHost->mpMap->mCustomEnvColors.contains(environmentId));
```

Comment:
A bit of code polishing... :grinning: 
```suggestion
            do {} while (mpHost->mpMap->mCustomEnvColors.contains(++environmentId));
```

Reply:
In this case I prefer the current version, it is easier to understand for those not deep in C++...

Reply:
Agree with @vadi2 im this case. Again, not even my code but taken verbatim from `T2DMap::slot_defineNewColor()`